File: cs/vmtest/SimpleTests.cs
namespace vmtest;

[TestClass]
public class SimpleTests
{
    [TestMethod]
    public void TestTest()
    {
        Assert.AreEqual(2, 1 + 1);
        //Assert.AreEqual(2, 1 - 1); // just to make sure tests can fail, just in case
    }

    [TestMethod]
    public void TestInstantiation()
    {
        VirtualMachine vm = new VirtualMachine();

        Assert.IsTrue(vm != null);
    }

    [TestMethod]
    public void TestNOP()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Execute(Bytecode.NOP);

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }

    [TestMethod]
    public void TestNOPs()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.NOP,
            Bytecode.NOP
        });

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }

    [TestMethod]
    public void TestPRINT()
    {
        VirtualMachine vm = new VirtualMachine();
        
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)12,
            Bytecode.PRINT
        });
        
        // Check out the output window to see if it printed

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }

    [TestMethod]
    public void TestDUMP()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Execute(new Bytecode[] {
            Bytecode.NOP,
            Bytecode.DUMP
        });

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }
    
    [TestMethod]
    public void TestTRACE()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Execute(new Bytecode[] {
            Bytecode.TRACE,
            Bytecode.NOP,
            Bytecode.DUMP
        });

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }
}

File: java/vm/src/test/java/simplevm/vm/SimpleTests.java
package simplevm.vm;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import static simplevm.vm.Bytecode.*;

class SimpleTests {
    @Test void testTest() {
        assertEquals(2, 1 + 1);
        //assertEquals(2, 1 - 1); // just to fail-test, just in case
    }
    @Test void testInstantiation() {
        VirtualMachine vm = new VirtualMachine();

        assertTrue(vm != null);
    }
    @Test void testNOP() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(NOP);

        // If we got here, with no exception, we're good
        assertTrue(vm != null);
    }
    @Test void testNOPs() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(new int[] {
            NOP,
            NOP,
            NOP,
            NOP
        });

        // If we got here, with no exception, we're good
        assertTrue(vm != null);
    }
    @Test void testDump() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(DUMP);

        // If we got here, with no exception, we're good
        assertTrue(vm != null);
    }
    @Test void testTrace() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(new int[] {
            TRACE,
            NOP,
            DUMP
        });

        // If we got here, with no exception, we're good
        assertTrue(vm != null);
    }
    @Test void testPrint() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(new int[] {
            CONST, 12,
            PRINT
        });

        // Check the output for the printed output

        // If we got here, with no exception, we're good
        assertTrue(vm != null);
    }
    @Test void testMoreDump() {
        VirtualMachine vm = new VirtualMachine();

        int[] code = {
            Bytecode.CONST, 43,
            Bytecode.NOP,
            Bytecode.DUMP
        };
        vm.execute(code);

        // If we got here, with no exception, we're good
        assertTrue(true);
    }
}
